drivers/staging/vc04_services/vc-sm-cma/vc_sm.c

```diff
@@ -237,14 +237,14 @@ static void vc_sm_add_resource(struct vc_sm_privdata_t *privdata,
 
 /*
  * Cleans up imported dmabuf.
+ * Should be called with mutex held.
  */
 static void vc_sm_clean_up_dmabuf(struct vc_sm_buffer *buffer)
 {
 	if (!buffer->imported)
 		return;
 
 	/* Handle cleaning up imported dmabufs */
-	mutex_lock(&buffer->lock);
```

Comment:
Was `vc_sm_dma_buf_release` never used before? I can't see how the double mutex_lock could have worked?

Reply:
Allocations by vcsm and imported dmabufs created dmabufs with different dma_buf_ops.

Allocs used ```vc_sm_dma_buf_release```, which called ```vc_sm_clean_up_dmabuf```, but as ```buffer->imported``` will have been 0, it immediately returned.

Imported dmabufs used ```vc_sm_import_dma_buf_release```, which omited to call ```vc_sm_clean_up_dmabuf```, hence the memory leak of the sgt due to not unmapping and detaching the buffer, but also never going through this path that would have double locked.

5.9 I'm ditching vcsm doing any allocations and getting userspace to handle the allocs via dma-heaps.